File: Baekjoon/Stack & Queue/균형잡힌 세상(4949).js
/**
 * 
 * 괄호의 균형이 맞는지 확인
 * () []
 * 
 * (,[ => ++
 * ),] => --
 * 
 */


const rl = require('readline').createInterface({
    input: process.stdin,
    output: process.stdout,
});

const input = [];

rl.on('line', (line) => {
    if (line === '.') rl.close();
    input.push(line);
});

rl.on('close', () => {
    for (const sentence of input) {
        const brackets = [];

        for (const character of sentence) {
            switch (character) {
                case '(':
                    brackets.push('(');
                    break;
                case ')':
                    if (brackets[brackets.length - 1 ] === '(') brackets.pop(')');
                    else brackets.push('!!');
                    break;
                case '[':
                    brackets.push('[');
                    break;
                case ']':
                    if (brackets[brackets.length - 1 ] === '[') brackets.pop(']');
                    else brackets.push('!!');
                    break;
                default:
                    break;
            }
        }

        console.log(brackets.length ? 'no' : 'yes');
    }
});

File: Baekjoon/Stack & Queue/균형잡힌 세상(4949).py
import sys
input = sys.stdin.readline

while True:
  sentence = input().rstrip()
  if sentence == '.':
    break

  brackets = []
  for character in sentence:
    if character == '(':
      brackets.append('(')
    elif character == '[':
      brackets.append('[')
    elif character == ')':
      if len(brackets) > 0 and brackets[len(brackets) - 1] == '(':
        brackets.pop()
      else: brackets.append('!')
    elif character == ']':
      if len(brackets) > 0 and  brackets[len(brackets) - 1] == '[':
        brackets.pop()
      else: brackets.append('!')
  
  print('no' if len(brackets) else 'yes')